benchmark/syclblas/expression/reduction_rows.cpp

```diff
@@ -0,0 +1,152 @@
+/**************************************************************************
+ *
+ *  @license
+ *  Copyright (C) 2016 Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename reduction_rows.cpp
+ *
+ **************************************************************************/
+
+#include "sycl_blas.hpp"
+#include "utils.hpp"
+
+using namespace blas;
+
+template <typename scalar_t>
+std::string get_name(int rows, int cols) {
+  std::ostringstream str{};
+  str << "BM_RedRows<" << blas_benchmark::utils::get_type_name<scalar_t>()
+      << ">/" << rows << "/" << cols;
+  return str.str();
+}
+
+template <typename operator_t, typename scalar_t, typename executor_t,
+          typename input_t, typename output_t>
+std::vector<cl::sycl::event> launch_reduction(executor_t& ex, input_t buffer_in, output_t buffer_out,
+                      index_t rows, index_t cols) {
+  blas::Reduction<operator_t, input_t, output_t, 64, 64, 4, scalar_t,
+                  static_cast<int>(Reduction_t::partial_rows)>
+      reduction(buffer_in, buffer_out, rows, cols);
+  return ex.execute(reduction);
+}
+
+template <typename scalar_t>
+void run(benchmark::State& state, ExecutorType* executorPtr, index_t rows,
+         index_t cols, bool* success) {
+  // The counters are double. We convert m, n and k to double to avoid integer
+  // overflows for n_fl_ops and bytes_processed
+  double rows_d = static_cast<double>(rows);
```

Comment:
Is it possible to use an unsigned integer type of suitable width here, for example uint64_t? Double seems an odd choice when floating point is not required.

Edit: Also I don't know how relevant it is but perhaps useful on platforms which don't support doubles? Had an error about this on rcar boards today which brought it to mind but again not sure how relevant a problem that is, or if platforms that don't support doubles are likely to support uint64_t.

Reply:
I know this is odd, but the Google Benchmark framework only supports `double` for the counters. I could have used a static cast to a `uint64_t` and then when assigning the counters there would be an implicit cast to `double`, but doing it that way seems better to me because it explicitly shows that the counters are double.

Concerning the platforms that don't support doubles, it's just that doubles have the size of a float. Using floats for the calculation of the counters is fine, the precision is good enough.

Reply:
And also, just wondering: on RCar does the CPU not support doubles? I think it's only the accelerator.

Reply:
Ah OK, that makes sense then! I'm not sure about the RCar CPU as I'd just seen the error in some of my own work, but you are probably right it is just the accelerator rather than the CPU. I'll resolve this now then :)